pkg/operator/ozone/controller.go

```diff
@@ -0,0 +1,280 @@
+		case "StatefulSet":
+			err = c.applyStatefulSet(resource, objectStore)
+			if err != nil {
+				logger.Errorf("Can't apply StatefulSet %s %s ", resource.Name(), err.Error())
+			}
+		case "Service":
+			err = c.applyService(resource, objectStore)
+			if err != nil {
+				logger.Errorf("Can't apply Service %s %s ", resource.Name(), err.Error())
+			}
+		case "ConfigMap":
+			err = c.applyConfigMap(resource, objectStore)
+			if err != nil {
+				logger.Errorf("Can't apply ConfigMap %s %s ", resource.Name(), err.Error())
+			}
+		default:
+			logger.Errorf("Unhandled resource type: %s", resource.Kind())
+		}
+	}
+	return nil
+}
+
+//helper function to set the owner for k8s resoures
+func (c *Controller) setOwnerRef(objectStore *ozonev1alpha1.ObjectStore, destination *metav1.ObjectMeta) {
+	ownerRef := metav1.OwnerReference{
+		APIVersion: ObjectStoreResource.Version,
+		Kind:       ObjectStoreResource.Kind,
+		Name:       objectStore.Namespace,
+		UID:        types.UID(objectStore.ObjectMeta.UID),
+	}
+	k8sutil.SetOwnerRef(destination, &ownerRef)
+}
+
+//parse real kubernetes object from a generated resource
+func (c *Controller) parseRenderedResource(resource *data.Resource,
+	destinationPointer interface{}) error {
+	resourceContent, err := resource.Content.ToString()
+	if err != nil {
+		return err
+	}
+
+	err = yaml.NewYAMLOrJSONDecoder(bytes.NewBufferString(resourceContent), 1000).Decode(destinationPointer)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//Method to be called in case of the CRD is added
+func (c *Controller) onAdd(obj interface{}) {
+	objectstore := obj.(*ozonev1alpha1.ObjectStore).DeepCopy()
+
+	logger.Infof("Ozone object store object is added/updated  %s", objectstore.Name)
+
+	err := c.generateResources(objectstore)
+	if err != nil {
+		logger.Errorf("Operation is failed for object store %s %s", objectstore.Name, err.Error())
+	}
+
+}
+
+//Method to be called in case of the CRD is updatede
+func (c *Controller) onUpdate(oldObj, newObj interface{}) {
+	newStore := newObj.(*ozonev1alpha1.ObjectStore).DeepCopy()
+	c.onAdd(newStore)
+}
+
+//Method to be called in case of the CRD is deleted
+func (c *Controller) onDelete(obj interface{}) {
+	objectstore := obj.(*ozonev1alpha1.ObjectStore).DeepCopy()
+	logger.Infof("Delete Ozone object store %s", objectstore.Name)
+	// Cleanup is handled by the owner references set in 'onAdd' and the k8s garbage collector.
+}
+
+//helper function to generate and deploy Service after the custom transformations
+func (c *Controller) applyService(resource *data.Resource,
+	objectStore *ozonev1alpha1.ObjectStore) error {
+	logger.Infof("Applying Service %s", resource.Name())
+	service := corev1.Service{}
+	err := c.parseRenderedResource(resource, &service)
+	if err != nil {
+		return err
+	}
+	c.setOwnerRef(objectStore, &service.ObjectMeta)
+	original, err := c.context.Clientset.CoreV1().Services(objectStore.Namespace).Get(resource.Name(), metav1.GetOptions{})
```

Comment:
There is a helper `k8sutil.CreateOrUpdateService()` that you could call. There may also be helpers in that package for configmaps and statefulsets, or else you could add those helpers.

Reply:
Thanks the hint. I started to use the helper for the service but no helper for daemonset/statefulset. Will create more in follow up PRs.